File: src/data/courses/dsa/arrays-and-hashing/array-manipulation/next-permutation/code/java/NextPermutation.java
// ============================================================
// Next Permutation
// ============================================================

// ============================================================
// Solution: Three-Step Algorithm (Optimal)
// Time: O(N) | Space: O(1)
// ============================================================
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length, i = n - 2;

        // Step 1: Find rightmost dip
        while (i >= 0 && nums[i] >= nums[i + 1]) i--;

        if (i >= 0) {
            // Step 2: Find rightmost element > nums[i]
            int j = n - 1;
            while (nums[j] <= nums[i]) j--;

            // Step 3: Swap
            int temp = nums[i]; nums[i] = nums[j]; nums[j] = temp;
        }

        // Step 4: Reverse suffix
        int l = i + 1, r = n - 1;
        while (l < r) {
            int temp = nums[l]; nums[l] = nums[r]; nums[r] = temp;
            l++; r--;
        }
    }
}

File: src/data/courses/dsa/arrays-and-hashing/array-manipulation/next-permutation/code/python/next_permutation.py
# ============================================================
# Next Permutation
# ============================================================

from typing import List

# ============================================================
# Solution: Three-Step Algorithm (Optimal)
# Time: O(N) | Space: O(1)
# ============================================================
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        n = len(nums)
        i = n - 2

        # Step 1: Find rightmost dip (nums[i] < nums[i+1])
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1

        if i >= 0:
            # Step 2: Find rightmost element > nums[i]
            j = n - 1
            while nums[j] <= nums[i]:
                j -= 1

            # Step 3: Swap
            nums[i], nums[j] = nums[j], nums[i]

        # Step 4: Reverse suffix [i+1:]
        nums[i + 1:] = reversed(nums[i + 1:])
